File: backend/app/schemas/social.py
from datetime import date

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.social import ChallengeMetric


class ChallengeCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=500)
    metric: ChallengeMetric
    starts_on: date
    ends_on: date

    @model_validator(mode="after")
    def _check_dates(self) -> "ChallengeCreate":
        if self.ends_on < self.starts_on:
            raise ValueError("ends_on no puede ser anterior a starts_on")
        return self


class ChallengeJoin(BaseModel):
    invite_code: str = Field(min_length=1, max_length=8)


class LeaderboardEntry(BaseModel):
    user_id: int
    name: str
    value: float
    rank: int
    is_me: bool


class ChallengeSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    metric: ChallengeMetric
    starts_on: date
    ends_on: date
    invite_code: str
    participant_count: int
    is_owner: bool


class ChallengeDetail(ChallengeSummary):
    description: str | None
    leaderboard: list[LeaderboardEntry]

File: lib/models/social.dart
class LeaderboardEntry {
  final String userId;
  final String name;
  final double value;
  final int rank;
  final bool isMe;

  LeaderboardEntry({
    required this.userId,
    required this.name,
    required this.value,
    required this.rank,
    required this.isMe,
  });
}

class ChallengeSummary {
  final String id;
  final String name;
  final String metric;
  final DateTime startsOn;
  final DateTime endsOn;
  final String inviteCode;
  final int participantCount;
  final bool isOwner;

  ChallengeSummary({
    required this.id,
    required this.name,
    required this.metric,
    required this.startsOn,
    required this.endsOn,
    required this.inviteCode,
    required this.participantCount,
    required this.isOwner,
  });
}

class ChallengeDetail extends ChallengeSummary {
  final String? description;
  final List<LeaderboardEntry> leaderboard;

  ChallengeDetail({
    required super.id,
    required super.name,
    required super.metric,
    required super.startsOn,
    required super.endsOn,
    required super.inviteCode,
    required super.participantCount,
    required super.isOwner,
    required this.description,
    required this.leaderboard,
  });
}

/// Metricas disponibles al crear un reto (value backend -> etiqueta ES + unidad).
const challengeMetrics = <String, String>{
  'total_volume_kg': 'Volumen total (kg)',
  'total_sessions': 'Sesiones entrenadas',
  'total_reps': 'Repeticiones totales',
};

String challengeMetricUnit(String metric) => switch (metric) {
  'total_volume_kg' => 'kg',
  'total_sessions' => 'sesiones',
  'total_reps' => 'reps',
  _ => '',
};
